### ISSUE TITLE
After installing latest forcedroid (version 4.1.2) on application launch get 

### ISSUE THREAD

**Issue by asinha08** [forcedotcom/SalesforceMobileSDK-Android#1118]
After installing latest forcedroid (version 4.1.2) I have added my client id and callbackurl. I am getting following error. Default application should list my users. But it through error like **com.salesforce.androidsdk.phonegap.plugin.SalesforceOAuthPlugin class not found**.

I also tried compelling the default application on android studio.

``
05-09 17:25:02.952 29166-29305/com.abc.xyz W/System.err: java.lang.ClassNotFoundException: com.salesforce.androidsdk.phonegap.plugin.SalesforceOAuthPlugin
05-09 17:25:02.952 29166-29305/com.abc.xyz W/System.err:     at java.lang.Class.classForName(Native Method)
05-09 17:25:02.952 29166-29305/com.abc.xyz W/System.err:     at java.lang.Class.forName(Class.java:324)
05-09 17:25:02.953 29166-29305/com.abc.xyz W/System.err:     at java.lang.Class.forName(Class.java:285)
05-09 17:25:02.953 29166-29305/com.abc.xyz W/System.err:     at org.apache.cordova.PluginManager.instantiatePlugin(PluginManager.java:489)
05-09 17:25:02.953 29166-29305/com.abc.xyz W/System.err:     at org.apache.cordova.PluginManager.getPlugin(PluginManager.java:169)
05-09 17:25:02.953 29166-29305/com.abc.xyz W/System.err:     at org.apache.cordova.PluginManager.exec(PluginManager.java:122)
05-09 17:25:02.953 29166-29305/com.abc.xyz W/System.err:     at org.apache.cordova.CordovaBridge.jsExec(CordovaBridge.java:59)
05-09 17:25:02.953 29166-29305/com.abc.xyz W/System.err:     at org.apache.cordova.engine.SystemExposedJsApi.exec(SystemExposedJsApi.java:41)
05-09 17:25:02.953 29166-29305/com.abc.xyz W/System.err:     at org.chromium.base.SystemMessageHandler.nativeDoRunLoopOnce(Native Method)
05-09 17:25:02.953 29166-29305/com.abc.xyz W/System.err:     at org.chromium.base.SystemMessageHandler.handleMessage(SystemMessageHandler.java:39)
05-09 17:25:02.953 29166-29305/com.abc.xyz W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:102)
05-09 17:25:02.953 29166-29305/com.abc.xyz W/System.err:     at android.os.Looper.loop(Looper.java:148)
05-09 17:25:02.953 29166-29305/com.abc.xyz W/System.err:     at android.os.HandlerThread.run(HandlerThread.java:61)
05-09 17:25:02.953 29166-29305/com.abc.xyz W/System.err: Caused by: java.lang.ClassNotFoundException: Didn't find class "com.salesforce.androidsdk.phonegap.plugin.SalesforceOAuthPlugin" on path: DexPathList[[zip file "/data/app/com.abc.xyz-2/base.apk"],nativeLibraryDirectories=[/data/app/com.abc.xyz-2/lib/arm, /vendor/lib, /system/lib]]
05-09 17:25:02.953 29166-29305/com.abc.xyz W/System.err:     at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
05-09 17:25:02.953 29166-29305/com.abc.xyz W/System.err:     at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
05-09 17:25:02.953 29166-29305/com.abc.xyz W/System.err:     at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
05-09 17:25:02.953 29166-29305/com.abc.xyz W/System.err: ... 13 more
05-09 17:25:02.953 29166-29305/com.abc.xyz W/System.err: Suppressed: java.lang.ClassNotFoundException: com.salesforce.androidsdk.phonegap.plugin.SalesforceOAuthPlugin
05-09 17:25:02.954 29166-29305/com.abc.xyz W/System.err:     at java.lang.Class.classForName(Native Method)
05-09 17:25:02.954 29166-29305/com.abc.xyz W/System.err:     at java.lang.BootClassLoader.findClass(ClassLoader.java:781)
05-09 17:25:02.954 29166-29305/com.abc.xyz W/System.err:     at java.lang.BootClassLoader.loadClass(ClassLoader.java:841)
05-09 17:25:02.954 29166-29305/com.abc.xyz W/System.err:     at java.lang.ClassLoader.loadClass(ClassLoader.java:504)
05-09 17:25:02.954 29166-29305/com.abc.xyz W/System.err: ... 14 more
05-09 17:25:02.954 29166-29305/com.abc.xyz W/System.err: Caused by: java.lang.NoClassDefFoundError: Class not found using the boot class loader; no stack trace available
05-09 17:25:02.954 29166-29305/com.abc.xyz I/System.out: Error adding plugin com.salesforce.androidsdk.phonegap.plugin.SalesforceOAuthPlugin.
05-09 17:25:02.954 29166-29305/com.abc.xyz W/System.err: java.lang.NullPointerException: Attempt to invoke virtual method 'void org.apache.cordova.CordovaPlugin.privateInitialize(java.lang.String, org.apache.cordova.CordovaInterface, org.apache.cordova.CordovaWebView, org.apache.cordova.CordovaPreferences)' on a null object reference
05-09 17:25:02.954 29166-29305/com.abc.xyz W/System.err:     at org.apache.cordova.PluginManager.getPlugin(PluginManager.java:171)
05-09 17:25:02.954 29166-29305/com.abc.xyz W/System.err:     at org.apache.cordova.PluginManager.exec(PluginManager.java:122)
05-09 17:25:02.954 29166-29305/com.abc.xyz W/System.err:     at org.apache.cordova.CordovaBridge.jsExec(CordovaBridge.java:59)
05-09 17:25:02.954 29166-29305/com.abc.xyz W/System.err:     at org.apache.cordova.engine.SystemExposedJsApi.exec(SystemExposedJsApi.java:41)
05-09 17:25:02.954 29166-29305/com.abc.xyz W/System.err:     at org.chromium.base.SystemMessageHandler.nativeDoRunLoopOnce(Native Method)
05-09 17:25:02.954 29166-29305/com.abc.xyz W/System.err:     at org.chromium.base.SystemMessageHandler.handleMessage(SystemMessageHandler.java:39)
05-09 17:25:02.955 29166-29305/com.abc.xyz W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:102)
05-09 17:25:02.955 29166-29305/com.abc.xyz W/System.err:     at android.os.Looper.loop(Looper.java:148)
05-09 17:25:02.955 29166-29305/com.abc.xyz W/System.err:     at android.os.HandlerThread.run(HandlerThread.java:61)
05-09 17:25:12.298 29166-29203/com.abc.xyz I/GMPM: Tag Manager is not found and thus will not be used
``

**Comment by asinha08** [comment]
Can anyone please comment on this issue..?

**Comment by abhinavguptas** [comment]
I am getting the same issue, with a basic hello world app created out of "forcedroid create"

`05-13 15:15:00.104 590-655/? D/ConnectivityService: releasing NetworkRequest NetworkRequest [ id=51, legacyType=-1, [ Capabilities: INTERNET&NOT_RESTRICTED&TRUSTED&NOT_VPN] ]
05-13 15:15:00.605 11031-11100/com.ccg24 W/System.err: java.lang.ClassNotFoundException: com.salesforce.androidsdk.phonegap.plugin.SalesforceOAuthPlugin
05-13 15:15:00.605 11031-11100/com.ccg24 W/System.err:     at java.lang.Class.classForName(Native Method)
05-13 15:15:00.605 11031-11100/com.ccg24 W/System.err:     at java.lang.Class.forName(Class.java:324)
05-13 15:15:00.605 11031-11100/com.ccg24 W/System.err:     at java.lang.Class.forName(Class.java:285)
05-13 15:15:00.605 11031-11100/com.ccg24 W/System.err:     at org.apache.cordova.PluginManager.instantiatePlugin(PluginManager.java:489)
05-13 15:15:00.605 11031-11100/com.ccg24 W/System.err:     at org.apache.cordova.PluginManager.getPlugin(PluginManager.java:169)`

**Comment by bhariharan** [comment]
@asinha08 It looks like the post install steps didn't get run to move the libraries to their correct locations. This typically happens when ```shelljs@0.5.3``` is not installed before running ```forcedroid```. Could you post your console output for the app creation process?

**Comment by meckert-salesforce** [comment]
I am/was having the same issue and can confirm that it's related to shelljs; apperently shelljs is declared a dependency for forcedroid, but when the `postinstall-android.js` is run it has a different search path for node modules and does not find it.  This is the same, wether I do a `forcedroid create` or a `cordova platform add android@5.0.0` (in an app created with forceios etc.).  Log with the `The node package shelljs is required to use this script. Run 'npm install shelljs@0.5.3' before running this script.` error: 
[1_log_no_shelljs.txt](https://github.com/forcedotcom/SalesforceMobileSDK-Android/files/272044/1_log_no_shelljs.txt)

When I do a `sudo npm install -g shelljs@0.5.3` and set the NODE_PATH environment variable to to the root of your node modules (e.g. `export NODE_PATH=/usr/local/lib/node_modules/`), `forcedroid create` runs without error: [2_log_with_shelljs.txt](https://github.com/forcedotcom/SalesforceMobileSDK-Android/files/272061/2_log_with_shelljs.txt)


So far so good, unfortunately once this goes through, I seem to be running into issue #1074

**Comment by asinha08** [comment]
@bhariharan, forcedroid installed on my machine has shelljs@0.5.3

**Comment by abhinavguptas** [comment]
Thanks for picking investigation on this issue, @huminzhi  @bhariharan, me, my team, and I know other teams, who nearly wasted more than 1 week in finding fix for this issue. We tried rolling back to 2014 versions of SDK to get ANT based builds etc in play, but nothing worked for us. By end of day, we are mostly Salesforce developers, who know Angular/Ionic and want to quickly give hybrid apps to our customers. Getting this deep into troubleshooting Cordova/Android builds becomes tricky for us. 


Thanks again, and hope to hear soon on the fix.

**Comment by bhariharan** [comment]
@asinha08 Could you please post your console output from ```forcedroid create```? The issue here is not just installation of ```shelljs@0.5.3```, but ensuring it's installed in the right path - local (vs) global.

**Comment by abhinavguptas** [comment]
@bhariharan sorry for jumping in, please find the "forcedroid create" console output below:

- **Cordova** Version : 6.1.1
- **Forcedroid** version 4.1.2

**Console Output**:
```
Abhinavs-MacBook-Pro:temp abhinav$ forcedroid create
Enter your application type (native, react_native, hybrid_remote, or hybrid_local): hybrid_local
Enter your application name: bug-demo
Enter the target directory of your app (must be an existing empty folder): .
Enter the package name for your app (com.mycompany.my_app): io.concret
Creating a new cordova project.
/Users/abhinav/code/temp/bug-demo /Users/abhinav/code/temp
Adding android project...
Creating Cordova project for the Android platform:
	Path: platforms/android
	Package: io.concret
	Name: bug_demo
	Activity: MainActivity
	Android target: android-23
Android project created with cordova-android@5.0.0
Discovered plugin "cordova-plugin-whitelist" in config.xml. Installing to the project
Fetching plugin "cordova-plugin-whitelist@1" via npm
Installing "cordova-plugin-whitelist" for android

               This plugin is only applicable for versions of cordova-android greater than 4.0. If you have a previous platform version, you do *not* need this plugin since the whitelist will be built in.
          
Fetching plugin "https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin" via git clone
Repository "https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin" checked out to git ref "v4.1.2".
Installing "com.salesforce" for android
Dependent plugin "cordova-plugin-whitelist" already installed on android.
Fetching plugin "https://github.com/apache/cordova-plugin-device" via git clone
Repository "https://github.com/apache/cordova-plugin-device" checked out to git ref "r1.0.1".
Installing "cordova-plugin-device" for android
Fetching plugin "https://github.com/phonegap/phonegap-plugin-push" via git clone
Repository "https://github.com/phonegap/phonegap-plugin-push" checked out to git ref "1.5.0".
Installing "phonegap-plugin-push" for android
Running SalesforceMobileSDK plugin android post-install script
The node package shelljs is required to use this script. Run 'npm install shelljs@0.5.3' before running this script.
/Users/abhinav/code/temp

Your application project is ready in ..

To use your new application in Android Studio, do the following:
   - Launch Android Studio and select `Import project (Eclipse ADT, Gradle, etc.)` from the Welcome screen
   - Navigate to the ./bug-demo/platforms/android folder, select it and click `Ok`
   - From the dropdown that displays the available targets, choose the sample app you want to run and click the play button

Before you ship, make sure to plug your OAuth Client ID,
Callback URI, and OAuth Scopes into www/bootconfig.json
Abhinavs-MacBook-Pro:temp abhinav$ 
```

**Comment by bhariharan** [comment]
@abhinavguptas Your console output clearly indicates that ```shelljs``` is not installed.

```
Running SalesforceMobileSDK plugin android post-install script
The node package shelljs is required to use this script. Run 'npm install shelljs@0.5.3' before running this script.
```

```forcedroid``` will not be able to perform the required post-install steps without ```shelljs```. You'll need to install ```shelljs@0.5.3``` in the context that ```forcedroid``` is being used before running the ```create``` command (local (vs) global).

**Comment by bhariharan** [comment]
Your console output clearly indicates that ```shelljs``` is not installed.

```
Running SalesforceMobileSDK plugin android post-install script
The node package shelljs is required to use this script. Run 'npm install shelljs@0.5.3' before running this script.
/Users/abhinav/code/temp
```

```forcedroid``` requires ```shelljs@0.5.3``` to be installed in the context it's being run in (local (vs) global), before the ```create``` command is run. Installing ```shelljs``` correctly should fix the issues you're facing.

**Comment by abhinavguptas** [comment]
I tried installing that as well @bhariharan, but nothing changes for me :( will try again and let you know

**Comment by bhariharan** [comment]
@abhinavguptas Did you install it locally or globally? The following steps should work:

* Create a new local directory and navigate to it.
* ```npm install forcedroid``` inside the new folder.
* ```npm install shelljs@0.5.3``` inside the new folder.
* ```node_modules/.bin/forcedroid create``` inside the new folder.
* The creation process kicks off and you should be able to import the generated project into ```Android Studio``` and run it.

**Comment by abhinavguptas** [comment]
Thanks @bhariharan, on both global and local forcedroid and shelljs install, I am getting this error. Following exactly the same steps followed by **$ cordova run android**

```
:processDebugManifest
/Users/abhinav/code/temp/newtest/platforms/android/AndroidManifest.xml:30:9-104 Error:
	Element activity#com.adobe.phonegap.push.PushHandlerActivity at AndroidManifest.xml:30:9-104 duplicated with element declared at AndroidManifest.xml:6:9-135
/Users/abhinav/code/temp/newtest/platforms/android/AndroidManifest.xml Error:
	Validation failed, exiting
:processDebugManifest FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':processDebugManifest'.

```

On checking the AndroidManifest.xml found that activity "com.adobe.phonegap.push.PushHandlerActivity" is added twice, snipped below:

```
<activity android:exported="true" android:launchMode="singleTop" android:name="com.adobe.phonegap.push.PushHandlerActivity" />
...
...
<activity android:configChanges="orientation|keyboardHidden|keyboard|screenSize|locale" android:label="@string/app_name" android:name="com.salesforce.androidsdk.phonegap.ui.SalesforceDroidGapActivity" android:theme="@android:style/Theme.Black.NoTitleBar">
    <intent-filter>
        <action android:name="android.intent.action.MAIN" />
        <category android:name="android.intent.category.LAUNCHER" />
    </intent-filter>
</activity>
<activity android:exported="true" android:name="com.adobe.phonegap.push.PushHandlerActivity" />

```

On removing one of the two duplicate activity, ran into following error:

```
:processDebugResources
/Users/abhinav/code/temp/newtest/platforms/android/build/intermediates/manifests/full/debug/AndroidManifest.xml:35:23-41 : No resource found that matches the given name (at 'icon' with value '@drawable/sf__icon').

:processDebugResources FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':processDebugResources'.
> com.android.ide.common.process.ProcessException: org.gradle.process.internal.ExecException: Process 'command '/Users/abhinav/tools/android-sdk-macosx/build-tools/23.0.1/aapt'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

```

**Comment by bhariharan** [comment]
@abhinavguptas Our instructions clearly specify not to use ```cordova build``` or ```cordova run```. Please use ```Android Studio``` or ```gradle``` directly. Also, the ```Salesforce``` plugin must be added last to avoid the duplicate entries error (if you have other plugins installed). Same goes for the icon error.

**Comment by abhinavguptas** [comment]
@bhariharan  I can't find a doc which says "NO" to Cordova build, which is essentially cordova compile + prepare, and the same is listed thru out the official Salesforce Mobile SDK Guide for hybrid apps, link https://developer.salesforce.com/docs/atlas.en-us.mobile_sdk.meta/mobile_sdk/hybrid_run_android.htm
Android studio is listed there as alternative way to run apps on Android. Its becoming quite a learning to make a hello world app run via official Mobile SDK.

I am working on Salesforce since 2008, and huge fan of this platform. But such tooling support, gap in documentation, and big struggle to execute a "hello world" with SDK is quite depressing. I am using Ionic in all other hybrid projects, and love the way "**$ ionic start myApp tabs**" works in one shot, without any such challenges. Again hybrid app development is about HTML/JS programming using Sublime/Atom with CLI tools for build and run, dependency on a big IDE like Android Studio will be troublesome, I am sure you guys will find ways to make salesforce mobile sdk plugins work normally as other cordova plugins do.  

I will try again on Android Studio, and let you know.

**Comment by bhariharan** [comment]
@abhinavguptas The output at the end of ```forcedroid create``` has instructions on importing into ```Android Studio``` or using ```gradle``` for command line builds. Our plugin uses our native libraries, which ```Cordova``` doesn't support very well, unfortunately. Most typical ```Cordova``` plugins are a few source files and nothing more, which makes it simple to bundle. We have postinstall steps to move the required libraries in place. We will look at making this simpler in the future, but in the meanwhile, ```Android Studio``` should work fine out of the box.

**Comment by abhinavguptas** [comment]
Alrite, so Android Studio + above local npm installs of shelljs and forcedroid get the hello world going :) 

Thanks @bhariharan for help !

We are using Ionic as a base to start the hybrid project, do you think following command with local npm installs, and Android studio work ?

```
$ cordova plugin add https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin
```

**Comment by bhariharan** [comment]
As long as ```shelljs@0.5.3``` is installed, it should work.

**Comment by rwhitley** [comment]
Hi Abhinav,

Thanks for your comments on the Dev Guide. We will clean up the alternate
cordova compile/emulate steps that you mentioned, since that route has
become too fragile for us to support.

This Trailhead project might help you with Ionic + Mobile SDK:

https://developer.salesforce.com/trailhead/project/mobile-sdk-hybrid-apps/mobilesdk-hybrid-apps-04

-Richard

Sent from my iPhone

On May 25, 2016, at 7:45 PM, Abhinav Gupta <notifications@github.com> wrote:

Alrite, so Android Studio + above local npm installs of shelljs and
forcedroid get the hello world going :)

Thanks @bhariharan <https://github.com/bhariharan> for help !

We are using Ionic as a base to start the hybrid project, do you think
following command with local npm installs, and Android studio work ?

$ cordova plugin add
https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub
<https://github.com/forcedotcom/SalesforceMobileSDK-Android/issues/1118#issuecomment-221763754>

**Comment by abhinavguptas** [comment]
Thanks @rwhitley, its quite interesting now, how any new developer trying to use Salesforce Mobile SDK can approach the steps. We have many stale WIKI pages on developer.salesforce.com, and few other blogs as well. Sometimes, google ranking of these stale but old content goes higher than dev guides, and they create bigger challenge. I can help with creating such obsolete content list, and we can do cleanup sometime. 

On side, I tried adding plugin to a newly created Ionic project, and imported it to Android Studio. Got the following error: https://dl.dropboxusercontent.com/u/1162324/temp/Screen%20Shot%202016-05-26%20at%209.25.40%20AM.png

I think my team tried the trailhead module for Ionic + Mobile SDK, but its failing as well, I will try again with Android Studio and update here.

**Comment by amjadajmer** [comment]
@bhariharan I created a new project using Forcedroid version 4.2 and changed the client Id in bootconfig.json, after that I executed cordova prepare and tried to run the app using android studio But it fails. Got the following issue:

E:\Mobile\test3\platforms\android\AndroidManifest.xml:30:9-104 Error:
	Element activity#com.adobe.phonegap.push.PushHandlerActivity at AndroidManifest.xml:30:9-104 duplicated with element declared at AndroidManifest.xml:6:9-135
E:\Mobile\test3\platforms\android\AndroidManifest.xml Error:
	Validation failed, exiting

 FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':processDebugManifest'.
> Manifest merger failed with multiple errors, see logs

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 1 mins 56.2 secs
